### ISSUE TITLE
Sequential errors for multiple batch dimensions, even when all inner layers are compatible

### ISSUE THREAD

**Issue by LarsKue** [keras-team/keras#21158]
I work on a use-case where we do training on single-instance batches of shape `(batch_size, dim)` and inference on sets of samples of shape `(batch_size, set_size, dim)`. This works well when using plain `keras.Layer`s, but `keras.Sequential` does an over-eager verification of the input shape that errors even when all inner layers are compatible with this use-case:

```py3
import keras

net = keras.Sequential([
    keras.layers.Dense(64)
])

# works
net(keras.ops.zeros((32, 2)))

# works when using the plain dense layer
net._layers[1](keras.ops.zeros((32, 100, 2)))

# errors when using Sequential
net(keras.ops.zeros((32, 100, 2)))
```

The only work-around is currently to build the network manually with Ellipsis `...` as the batch size:

```py3
net.build((..., 2))
```

It would be nice if the check could be relaxed, or dropped entirely.

**Comment by sonali-kumari1** [comment]
Hi @LarsKue - 
When you call `net(keras.ops.zeros((32, 2)))` for the first time, it will build the model with the input shape `(32, 2)`. However, when you later call `net(keras.ops.zeros((32, 100, 2)))`, it raises `ValueError` because once a model is built with a specific input shape(`(32, 2)` in your case), it expects all the succeeding inputs to have the same shape. Please refer to this documentation for more details on how [sequential model](https://keras.io/api/models/sequential/#sequential-class) works. I am also attaching a [gist](https://colab.research.google.com/gist/sonali-kumari1/e54d4a1dbfc609cf1674c6481dbe6fb8/21158.ipynb) for your reference. Thanks!

**Comment by LarsKue** [comment]
@sonali-kumari1 I understand this. What I am saying is that this behavior should not be intended; it is a bug.

Calling `Dense` builds it only on the last dimension of the input, as it should work. Why does `Sequential` even make an eager input shape check instead of letting the forward pass fail lazily from inner layers when the shape is incompatible?

The only benefit I can see is improved error messages, but this currently comes at the cost of a confusing and frankly limiting design.

**Comment by hertschuh** [comment]
@LarsKue 

Thank you for the report.

Keras can handle dynamic dimensions, but dynamic ranks are problematic in general. The recommended way to do what you're trying to do is:

```python
net(keras.ops.zeros((32, 1, 2)))

net(keras.ops.zeros((32, 100, 2)))
```
